png-read-write.md

```diff
@@ -0,0 +1,105 @@
+```d
+struct PngImage {
+    ...
+    // They return false on error but also update the status and error string in the structure
+    bool loadData(ref ubyte[] file, PngFormat desiredFormat = PngFormat.rgba);
+    bool loadData(ref ubyte[] file, Allocator* allocator, PngFormat desiredFormat = PngFormat.rgba) nothrow @nogc
+}
+```
+
+
+## Open Questions
+
+1. Supporting interlaced (progressive) format would bring in a lot of complexity, would probably require more API changes and it is very rare these days. Would we be ok with the decision not to support it until further notice? (Disclaimer: I am not implementing it either way :))
+2. Should we make a separate float gamma field in PngImage struct that gets populated from gAMA chunk automatically instead of having to look for it in chunk data linked list?
+3. Should we even use gamma from the file as given? This [article](https://hsivonen.fi/png-gamma/) speaks of problems with it.
+4. Should we add pixelFormat field that says in which format pixels array is, or is it enough that the caller knows what format he asked for?
+5. Should the package provide API with File and be made to work by loading part of the file at a time, instead of loading the whole file in memory before processing? This would add a solid amount of complexity...
+6. Should PngImage struct save the reference to the allocator and provide a `free` method that frees all the data it allocated from it (pixels and chunks linked list)?
+7. I assume most of the time the caller will want to just keep the pixels array and discard everything else so by default we shouldn't even allocate and load chunks linked list, but how do we then allow user to say "I do want chunk data and I will take care of deallocating it when needed"? Maybe also add functions like loadDataWithAllChunks(...) to the API?
+8. I decided to use one mmap/VirtualAlloc call (through MMapAllocator) to allocate all the memory I will need for temporary processing while I load and process the data. Will this be good enough for all use cases?
+9. Having desiredFormat in the API minimizes allocations and transformations needed to come to the final result but do we want to provide the transformation methods separately instead, at the cost of having to allocate multiple times?
+10. One specific transformation requires some color space magic: rgb to gray. Should we:
+	1. make this package depend on color space stuff in order to provide this,
+    2. just put the code needed for this one transformation in this package directly or
+    3. just say that rgb to gray and gray to rgb transformations are not supported by this package and you need to do it using a higer level API like ImageBuffer? ImageBuffer will probably need to support this either way so we could use that to keep this package simpler.
```

Comment:
Wrong way round. ImageBuffer is the base. The image format goes on top of that.

Reply:
Absolutely, you should give a gray ImageBuffer in that case